horizon/startup/offline_mode.py

```diff
@@ -0,0 +1,95 @@
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.backends import default_backend
+from cryptography.fernet import Fernet
+from pydantic import ValidationError
+import base64
+import secrets
+
+from horizon.startup.schemas import RemoteConfig, RemoteConfigBackup
+from opal_common.logger import logger
+
+
+class OfflineModeManager:
+    """
+    A backup for the remote config, in case the sidecar can't fetch the remote config.
+    """
+
+    def __init__(self, backup_path: str, api_key: str):
+        self._backup_path: str = backup_path
+        self._api_key = api_key
+
+    def _derive_backup_key(self, salt: Optional[bytes] = None) -> Tuple[bytes, bytes]:
+        if salt is None:
+            salt = secrets.token_bytes(16)
+        else:
+            salt = base64.urlsafe_b64decode(salt)
+
+        hkdf = HKDF(
+            algorithm=hashes.SHA256(),
+            length=32,
+            salt=salt,
+            info=b"Sidecar's local remote-config backup Key",
+            backend=default_backend(),
+        )
+        # We don't bother extracting the actual cryptographic bytes from the API key (which has a urlsafe encoding + a prefix),
+        # The 512-bit entropy is still there, and HKDF's extract phase handles inputs of non-uniform randomness.
+        key_bytes = hkdf.derive(self._api_key.encode("utf-8"))
+        return base64.urlsafe_b64encode(key_bytes), base64.urlsafe_b64encode(salt)
+
+    def backup_config(self, remote_config: RemoteConfig):
+        logger.info(
+            "Backing up remote config to {path}",
+            path=self._backup_path,
+        )
+
+        enc_key, salt = self._derive_backup_key()
+
+        try:
+            with open(self._backup_path, "w") as f:
+                f.write(
+                    RemoteConfigBackup(
+                        enc_remote_config=Fernet(enc_key).encrypt(
+                            remote_config.json(ensure_ascii=False).encode()
+                        ),
+                        key_derivation_salt=salt,
+                    ).json(ensure_ascii=False)
+                )
+        except Exception as e:
+            logger.exception("Failed to backup sidecar config")
+
+    def restore_config(self) -> Optional[RemoteConfig]:
+        logger.info(
+            "Loading config from local backup at {path}",
+            path=self._backup_path,
+        )
+        remote_config_backup: RemoteConfigBackup
+        try:
+            with open(self._backup_path, "r") as f:
+                remote_config_backup = RemoteConfigBackup.parse_raw(f.read())
+        except FileNotFoundError:
+            logger.warning("Local backup file of sidecar config not found")
+            return None
+        except ValidationError:
+            logger.error("Failed to parse sidecar config backup file")
+            return None
+
+        dec_key, _ = self._derive_backup_key(remote_config_backup.key_derivation_salt)
```

Comment:
you store the encryption key together with the encrypted data ? doesn't it mean the encryption is redundant ?